flink-connector-cassandra/src/main/java/org/apache/flink/connector/cassandra/source/enumerator/CassandraSplitEnumerator.java

```diff
@@ -0,0 +1,156 @@
+    }
+
+    @Override
+    public void handleSplitRequest(int subtaskId, @Nullable String requesterHostname) {
+        assignUnprocessedSplitsToReader(subtaskId);
+    }
+
+    @Override
+    public void start() {
+        // discover the splits and update unprocessed splits and then assign them.
+        // There is only an initial splits discovery, no periodic discovery.
+        enumeratorContext.callAsync(
+                this::discoverSplits,
+                (splits, throwable) -> {
+                    LOG.info("Add {} splits to CassandraSplitEnumerator.", splits.size());
+                    state.addNewSplits(splits, enumeratorContext.currentParallelism());
+                });
+    }
+
+    private List<CassandraSplit> discoverSplits() {
+        final int numberOfSplits = enumeratorContext.currentParallelism();
+        final Metadata clusterMetadata = cluster.getMetadata();
+        final String partitioner = clusterMetadata.getPartitioner();
+        final SplitsGenerator splitsGenerator = new SplitsGenerator(partitioner);
+        if (MURMUR3PARTITIONER.equals(partitioner)) {
+            LOG.info("Murmur3Partitioner detected, splitting");
+            List<BigInteger> tokens =
+                    clusterMetadata.getTokenRanges().stream()
+                            .map(
+                                    tokenRange ->
+                                            new BigInteger(
+                                                    tokenRange.getEnd().getValue().toString()))
+                            .collect(Collectors.toList());
+            return splitsGenerator.generateSplits(numberOfSplits, tokens);
+        } else {
+            // Murmur3Partitioner is the default and recommended partitioner for Cassandra 1.2+
+            // see
+            // https://docs.datastax.com/en/cassandra-oss/3.x/cassandra/architecture/archPartitionerAbout.html
+            LOG.warn(
+                    "The current Cassandra partitioner is {}, only Murmur3Partitioner is supported "
+                            + "for splitting, using an single split",
+                    partitioner);
+            return splitsGenerator.generateSplits(1, Collections.emptyList());
+        }
+    }
+
+    @Override
+    public void addSplitsBack(List<CassandraSplit> splits, int subtaskId) {
+        LOG.info("Add {} splits back to CassandraSplitEnumerator.", splits.size());
+        state.addNewSplits(splits, enumeratorContext.currentParallelism());
+        assignUnprocessedSplitsToReader(subtaskId);
+    }
+
+    @Override
+    public void addReader(int subtaskId) {
+        LOG.info("Adding reader {} to CassandraSplitEnumerator.", subtaskId);
+        assignUnprocessedSplitsToReader(subtaskId);
+    }
+
+    private void assignUnprocessedSplitsToReader(int readerId) {
+        checkReaderRegistered(readerId);
+
+        final Set<CassandraSplit> splitsForReader = state.getSplitsForReader(readerId);
```

Comment:
For something that is mostly targeted at batch workloads it seems strange to pre-assign splits to readers. Rather they should be fully distributed on demand to prevent stragglers.

Reply:
:+1:

Reply:
Still, I find something strange. If I don't call `assignUnprocessedSplitsToReader()` in `addReader()` splits never get assigned because`handleSplitRequest()` was never called by the source framework.

Reply:
It looks like the fetching of additional splits isn't implemented generically. The FileSourceReader explicitly requests requests whenever it has finished a split, and I don't see anything where this happens automatically.

Reply:
IMHO when I override a method such as `handleSplitRequest()` that a framework provides, I'm kind of expecting it to be called by the framework and not having to call it myself. The doc does not mention that we need to call it so I suggest we should open a ticket to improve the framework on this side. I'll open the ticket, feel free to comment it or close it if you disagree.

Reply:
https://issues.apache.org/jira/browse/FLINK-30805